starlette/middleware/base.py

```diff
@@ -24,19 +25,52 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
             await self.app(scope, receive, send)
             return
 
+        response_sent = anyio.Event()
+
         async def call_next(request: Request) -> Response:
             app_exc: typing.Optional[Exception] = None
             send_stream, recv_stream = anyio.create_memory_object_stream()
 
+            async def receive_or_disconnect() -> Message:
+                if response_sent.is_set():
+                    return {"type": "http.disconnect"}
+
+                async with anyio.create_task_group() as task_group:
+
+                    async def wrap(func: typing.Callable[[], typing.Awaitable[T]]) -> T:
+                        result = await func()
+                        task_group.cancel_scope.cancel()
+                        return result
+
+                    task_group.start_soon(wrap, response_sent.wait)
+                    message = await wrap(request.receive)
```

Comment:
What this is doing is saying "wait for a message from the client but if `response_sent` gets set in the meantime then stop waiting/reading from the client and move on"

Reply:
Yes, the issue that I want to solve here is, if the downstream app is waiting on `receive`, but as the response is sent (likely by the middleware), the downstream app cannot `send` anything meaningfully, so there is no point in letting downstream wait for another message from upstream.

Of course, we could also choose to rely on upstream `receive` returning a `http.disconnect` message when the response is sent (which *should* happen), but when I wrote that bit, I thought that a belt-and-braces approach would be better.

Reply:
In other words, I would be open to modifying that part to remove the `receive` wrapper to avoid that cost if it thought to be too much.